### DISCUSSION TITLE
Check out this build

### DISCUSSION THREAD

**Discussion by 38github** [mikeoliphant/neural-amp-modeler-lv2#61] [Show and tell]
https://didthis.app/user/4xvkqmpk/project/nvq5x/post/33yj6

**Comment by mikeoliphant** [comment]
I was unable to download anything from that link.

What kind of performance increase are you seeing relative to the current lv2 plugin build with USE_NATIVE_ARCH on?

**Comment by 38github** [comment]
This is what my Didthis post says:

http://coginthemachine.ddns.net/mnt/nam/software_src_misc/neural-amp-modeler-lv2_git20240326_linux_x86-64-v3_o3_lto_pgo_openacc_glibc239.tar.bz2

O3, LTO, PGO, OpenACC and x86-64-v3 optimized Linux build of neural-amp-modeler-lv2 for Glibc 2.39 systems.

IMPROVEMENT! On my Ryzen 5600G with a fixed 4.4GHz CPU frequency a Standard WaveNet NAM model used 0.40% CPU. The build from yesterday used 0.5%.

The following is not a system benchmark but just an anecdote. I did a test on my i3-7100U CPU running this build of neural-amp-modeler-lv2 on Linux Void and then on Windows 10. The plugin on Windows 10 was compiled using Clang 16 (O3, LTO and x86-64-v3). The DAW on both systems was REAPER and an LSTM model trained with 2 num_layers and 14 hidden_size. 

The Linux system managed up to 73 instances without any crackling or xruns.

The Windows system was unresponsive using only 26 instances. If I had used the official Neural Amp Modeler VST3 I am certain it would have been even fewer instances before unresponsiveness.

**Reply by mikeoliphant** [reply]
That link does not work for me.

So is it 0.40% vs 0.50%? If so, that is significant. If it is .4% vs .5% (only one significant digit) then you can't tell since it is being rounded to the nearest .1%

**Reply by 38github** [reply]
The 0.40 vs 0.50 is just from the build from yesterday. It was two decimals.
[neural-amp-modeler-lv2_git20240326_linux_x86-64-v3_o3_lto_pgo_openacc_glibc239.zip](https://github.com/mikeoliphant/neural-amp-modeler-lv2/files/14776876/neural-amp-modeler-lv2_git20240326_linux_x86-64-v3_o3_lto_pgo_openacc_glibc239.zip)

**Reply by mikeoliphant** [reply]
Can you link me a Windows version of that build?

**Reply by 38github** [reply]
I don't have a Windows version of it, unfortunately. I have to figure out how to do all the steps using Visual Studio and the rest. I am compiling an aarch64/arm64 that can be used on a Raspberry Pi using a Debian Bullseye based distro (Glibc 2.36)

**Reply by mikeoliphant** [reply]
If you get a build that runs better than the default build on Raspberry Pi, let me know - I'd be interested in that.

**Reply by 38github** [reply]
Wow. I can now run a WaveNet Standard model on my RPi3 kind of SBC where a feather model before used 90% CPU. The standard model now use 93% with this build and the feather model 48%. This is great.

[neural-amp-modeler-lv2_git20240326_linux_aarch64_o3_lto_pgo_openacc_glibc236.zip](https://github.com/mikeoliphant/neural-amp-modeler-lv2/files/14779198/neural-amp-modeler-lv2_git20240326_linux_aarch64_o3_lto_pgo_openacc_glibc236.zip)

**Comment by mikeoliphant** [comment]
If you can outline the changes to the build procedure, I can try to replicate it on my RPi4.

**Reply by 38github** [reply]
I will have to write down the steps because there are quite some compared to a regular build. I will have to get back to you.

**Comment by 38github** [comment]
I have typed down a guide:
```
######################################################
Build instruction for Debian 12/Bookworm arm64/aarch64
######################################################


Add LLVM/Clang 18 repository under su to file /etc/apt/sources.list.d/llvm18.list:
# echo -e "deb http://apt.llvm.org/bookworm/ llvm-toolchain-bookworm-18 main \ndeb-src http://apt.llvm.org/bookworm/ llvm-toolchain-bookworm-18 main" >> /etc/apt/sources.list.d/llvm18.list

Proceed to apt update and install necessary packages. E.g. clang-18 clang-tools-18 llvm-18 libclang-rt-18-dev 


Download neural-amp-modeler-lv2 source code and enter the build directory:
$ cd /tmp && git clone --recurse-submodules -j4 https://github.com/mikeoliphant/neural-amp-modeler-lv2 && cd neural-amp-modeler-lv2/build


Export Clang 18 flags/parameters (compile with -Og) and then compile:
$ export CC=clang-18 && export CXX=clang++-18 && export LD=lld-18 && export CFLAGS="-Og -fprofile-generate=/tmp/neural-amp-modeler-lv2/build" && export CXXFLAGS="-Og -fprofile-generate=/tmp/neural-amp-modeler-lv2/build" && export LDFLAGS="-Og -fprofile-generate=/tmp/neural-amp-modeler-lv2/build"
$ cmake .. && make


- Copy the created "neural_amp_modeler.lv2" folder to ~/.lv2/
- Run the plugin with your LV2 host of choice and try load several models with different activators, algorithms, settings, etc and adjust sliders randomly.
- Close the plugin/host

A .profraw file has been created in /tmp/neural-amp-modeler-lv2/build/


Go back to build directory and convert the .profraw data to "profile.profdata":
$ cd /tmp/neural-amp-modeler-lv2/build/
$ llvm-profdata-18 merge -output profile.profdata RENAMETOACTUALFILENAME.profraw


Remove all files in /tmp/neural-amp-modeler-lv2/build/ EXCEPT for profile.profdata .


Compile with optimized flags/parameters:
$ export CC=clang-18 && export CXX=clang++-18 && export LD=lld-18 && export CFLAGS="-O3 -flto -fopenacc -fprofile-use=/tmp/neural-amp-modeler-lv2/build/profile.profdata" && export CXXFLAGS="-O3 -flto -fopenacc -fprofile-use=/tmp/neural-amp-modeler-lv2/build/profile.profdata" && export LDFLAGS="-O3 -flto -fopenacc -fprofile-use=/tmp/neural-amp-modeler-lv2/build/profile.profdata"
$ cmake .. -DBUILD_SHARED_LIBS=0 -DCMAKE_BUILD_TYPE=Release && make

Move newly compiled optimized files and folder neural_amp_modeler.lv2 to directory ~/.lv2/ . Replace old files.

Now you should have the following files:

~/.lv2/neural_amp_modeler.lv2/manifest.ttl
~/.lv2/neural_amp_modeler.lv2/neural_amp_modeler.so
~/.lv2/neural_amp_modeler.lv2/neural_amp_modeler.ttl

// R Gerthsson
```

**Comment by 38github** [comment]
What are your thoughts?

**Reply by mikeoliphant** [reply]
The process is involved enough that I haven't had a chance to try it yet.

**Reply by 38github** [reply]
Have you tried the precompiled?

**Reply by mikeoliphant** [reply]
The only Linux system I've got is a Raspberry Pi, but it is running bullseye.

**Reply by 38github** [reply]
I set up a QEMU aarch64 VM and compiled a Bullseye build that you can try:
http://coginthemachine.ddns.net/mnt/nam/software_src_misc/neural-amp-modeler-lv2_git20240412_modified_linux_aarch64_o3_lto_pgo_openacc_glibc231.tar.bz2

**Reply by mikeoliphant** [reply]
Ok, that build works on my RPi4.

Using your build, standard NAM models perform slightly worse than a default build. I also tried a 1x12 LSTM model, and it *might* be a very slight bit faster with your build.

**Reply by 38github** [reply]
Worse? Interesting. Have you set your CPU frequency to a fixed value?

**Reply by mikeoliphant** [reply]
I haven't messed with any CPU settings.

It is possible that a virtual machine doesn't provide a great environment for doing PGO.

**Reply by 38github** [reply]
> It is possible that a virtual machine doesn't provide a great environment for doing PGO.

I use the same build on my RPi5 and it shows the same performance as the "non VM" build. Try executing "sudo cpupower frequency-set -g performance" to set it at the highest frequency.

**Comment by mikeoliphant** [comment]
> Try executing "sudo cpupower frequency-set -g performance" to set it at the highest frequency.

Makes no difference (which isn't surprising, since running NAM is more than enough to kick the core being used into full power).

**Reply by 38github** [reply]
Thank you for testing it. I will have to look into why we see different results. Regards.
